homeassistant/components/fritzbox/light.py

```diff
@@ -129,16 +132,34 @@ async def async_turn_on(self, **kwargs: Any) -> None:
             level = kwargs[ATTR_BRIGHTNESS]
             await self.hass.async_add_executor_job(self.device.set_level, level)
         if kwargs.get(ATTR_HS_COLOR) is not None:
-            hass_hue = int(kwargs[ATTR_HS_COLOR][0])
-            hass_saturation = round(kwargs[ATTR_HS_COLOR][1] * 255.0 / 100.0)
-            # find supported hs values closest to what user selected
-            hue = min(self._supported_hs.keys(), key=lambda x: abs(x - hass_hue))
-            saturation = min(
-                self._supported_hs[hue], key=lambda x: abs(x - hass_saturation)
-            )
-            await self.hass.async_add_executor_job(
-                self.device.set_color, (hue, saturation)
-            )
+            # Try setunmappedcolor first. This allows free color selection,
+            # but we don't know if its supported by all devices.
+            try:
+                # HA gives 0..360 for hue, fritz light only supports 0..359
+                unmapped_hue = int(kwargs[ATTR_HS_COLOR][0] % 360)
+                unmapped_saturation = round(kwargs[ATTR_HS_COLOR][1] * 255.0 / 100.0)
+                await self.hass.async_add_executor_job(
+                    self.device.set_unmapped_color, (unmapped_hue, unmapped_saturation)
+                )
+            # This will raise 400 BAD REQUEST if the setunmappedcolor is not available
+            except HTTPError as err:
+                if err.response.status_code == 400:
```

Comment:
I do not think we need to check for specific error code, since there was already no explicit error code check before - just do the fallback in case of `HTTPError`

Reply:
If there is a different error, the second request will most likely fail too, so I only did the fallback in case of error 400.

Reply:
in this case, I would suggest to raise a `HomeAssistantError` from the `HTTPError` so that a related message is shown in the UI to the user.
Further I would suggest something like:
```suggestion
                if err.response.status_code != 400:
                    raise HomeAssistantError(str(err)) from err

                # do the fallback
```
with this the fallback code block does not need to be indented

Reply:
If the error code is not 400 the original error gets passed on and handled by HA just like if there would not be any handling here. I think we should keep it that way for now, as we don't do any handling of `HTTPErrors` on other requests.

Swapped the if-clause to have less indentation.